patch/llpcPatchCopyShader.cpp

```diff
@@ -412,169 +404,262 @@ void PatchCopyShader::ExportOutput(
         auto pOutputTy = VectorType::get(m_pContext->FloatTy(), dwordSize);
         pOutputValue = UndefValue::get(pOutputTy);
 
-        for (uint32_t i = 0; i < dwordSize; ++i)
+#if LLPC_BUILD_GFX10
+        if (m_pContext->GetNggControl()->enableNgg)
         {
-            auto pLoadValue = LoadValueFromGsVsRingBuffer(loc + i / 4, i % 4, streamId, pInsertPos);
-            pOutputValue = InsertElementInst::Create(pOutputValue,
-                                                     pLoadValue,
-                                                     ConstantInt::get(m_pContext->Int32Ty(), i),
-                                                     "",
-                                                     pInsertPos);
+            // NOTE: For NGG, importing GS output from GS-VS ring is represented by a call and the call is replaced with
+            // real instructions when when NGG primitive shader is generated.
+            std::vector<Value*> args;
+            args.push_back(ConstantInt::get(m_pContext->Int32Ty(), loc));
+            args.push_back(ConstantInt::get(m_pContext->Int32Ty(), 0));
+            args.push_back(ConstantInt::get(m_pContext->Int32Ty(), streamId));
+
+            std::string callName = LlpcName::NggGsOutputImport + GetTypeName(pOutputTy);
+            pOutputValue = EmitCall(m_pModule, callName, pOutputTy, args, NoAttrib, pInsertPos);
+        }
+        else
+#endif
+        {
+            for (uint32_t i = 0; i < dwordSize; ++i)
+            {
+                auto pLoadValue = LoadValueFromGsVsRingBuffer(loc + i / 4, i % 4, streamId, pInsertPos);
+                pOutputValue = InsertElementInst::Create(pOutputValue,
+                                                         pLoadValue,
+                                                         ConstantInt::get(m_pContext->Int32Ty(), i),
+                                                         "",
+                                                         pInsertPos);
+            }
         }
         ExportGenericOutput(pOutputValue, loc, streamId, pInsertPos);
     }
 
-    pOutputValue = UndefValue::get(m_pContext->Floatx4Ty());
-    if (builtInUsage.position)
+#if LLPC_BUILD_GFX10
+    if (m_pContext->GetNggControl()->enableNgg)
     {
-        LLPC_ASSERT(pResUsage->inOutUsage.builtInOutputLocMap.find(BuiltInPosition) !=
-            pResUsage->inOutUsage.builtInOutputLocMap.end());
+        std::vector<std::pair<BuiltIn, Type*>> builtInPairs;
 
-        uint32_t loc = pResUsage->inOutUsage.builtInOutputLocMap[BuiltInPosition];
+        if (builtInUsage.position)
+        {
+            builtInPairs.push_back(std::make_pair(BuiltInPosition, m_pContext->Floatx4Ty()));
+        }
 
-        for (uint32_t i = 0; i < 4; ++i)
+        if (builtInUsage.pointSize)
```

Comment:
can we merge ngg and non-ngg path? the code looks similar.

Reply:
I can make another change to do this merge because original copy shader's processing built-ins are not well organized.